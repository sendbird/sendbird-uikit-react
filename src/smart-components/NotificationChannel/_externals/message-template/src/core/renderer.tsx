import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
