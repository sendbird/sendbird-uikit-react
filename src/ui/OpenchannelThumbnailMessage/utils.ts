export const SUPPORTING_TYPES = {
  IMAGE: 'IMAGE',
  VIDEO: 'VIDEO',
  UNSUPPORTED: 'UNSUPPORTED',
};

const SUPPORTED_MIMES = {
  IMAGE: [
    'image/jpeg',
    'image/jpg',
    'image/png',
    'image/gif',
    'image/webp',
  ],
  VIDEO: [
    'video/mpeg',
    'video/ogg',
    'video/webm',
    'video/mp4',
  ],
};

export const getSupportingFileType = (type: string): string => {
  if (SUPPORTED_MIMES.IMAGE.indexOf(type) >= 0) {
    return SUPPORTING_TYPES.IMAGE;
  }
  if (SUPPORTED_MIMES.VIDEO.indexOf(type) >= 0) {
    return SUPPORTING_TYPES.VIDEO;
  }
  return SUPPORTING_TYPES.UNSUPPORTED;
};
