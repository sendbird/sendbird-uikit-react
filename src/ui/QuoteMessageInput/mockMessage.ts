export const dummyFileMessageImage = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: 'https://static.sendbird.com/sample/user_sdk/user_sdk_16.png',
  name: 'Hello i am PDF',
  size: 0,
  type: 'image/pdf',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageVideo = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: 'https://sendbird-upload.s3.amazonaws.com/2D7B4CDB-932F-4082-9B09-A1153792DC8D/upload/n/fd4e65eb495f436ab687ba7c5683c922.mp4',
  name: 'https://sendbird-upload.s3.amazonaws.com/2D7B4CDB-932F-4082-9B09-A1153792DC8D/upload/n/fd4e65eb495f436ab687ba7c5683c922.mp4',
  size: 0,
  type: 'video/mp4',
  thumbnails: [{
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 240,
    height: 240
  }, {
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 320,
    height: 320
  }],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageAudio = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3',
  name: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3',
  size: 0,
  type: 'audio/midi',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0,
};

export const dummyFileMessageGif = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: 'https://static.sendbird.com/sample/user_sdk/user_sdk_16.png',
  name: 'GIF GIF GIF',
  size: 0,
  type: 'image/gif',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: '6e3053ba-c47b-47dc-bb9a-5814091f9553',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageWithThumbnail = {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  isFileMessage: () => true,
  messageId: 751865890,
  messageType: "file",
  channelUrl: "sendbird_group_channel_19238732_8e44958ac5c2fb88203146f42a18a9a4f4472dde",
  data: "",
  customType: "",
  createdAt: 1583104853295,
  updatedAt: 0,
  channelType: "group",
  metaArrays: [],
  reactions: [],
  mentionType: "users",
  mentionedUsers: [],
  url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
  name: "ic-close.svg",
  size: 404,
  type: "image/svg+xml",
  thumbnails: [{
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 240,
    height: 240
  }, {
    url: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    real_height: 20,
    real_width: 20,
    width: 320,
    height: 320
  }],
  sender: {
    nickname: "hoon1000",
    profileUrl: "https://static.sendbird.com/sample/user_sdk/user_sdk_16.png",
    userId: "hoon1000",
    connectionStatus: "nonavailable",
    lastSeenAt: 0,
    metaData: {},
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: "1583104837675",
  requireAuth: true,
  requestState: "succeeded",
  requestedMentionUserIds: [],
  errorCode: 0
};

export default { dummyFileMessageAudio, dummyFileMessageVideo, dummyFileMessageImage, dummyFileMessageWithThumbnail }
