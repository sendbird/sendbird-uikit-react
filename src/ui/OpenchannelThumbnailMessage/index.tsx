import React, {
  useRef,
  useMemo,
  useState,
  useEffect,
  useContext,
} from 'react';
import format from 'date-fns/format';
import './index.scss';
import { SUPPORTING_TYPES, getSupportingFileType } from './utils';
import { ClientFileMessage } from '../../index';
import { LocalizationContext } from '../../lib/LocalizationContext';

import Avatar from '../Avatar';
import ContextMenu, { MenuItems, MenuItem } from '../ContextMenu';
import Icon, { IconTypes, IconColors } from '../Icon';
import IconButton from '../IconButton';
import ImageRenderer from '../ImageRenderer';
import Label, { LabelTypography, LabelColors } from '../Label';
import Loader from '../Loader';
import UserProfile from '../UserProfile';
import { UserProfileContext } from '../../lib/UserProfileContext';
import {
  checkIsByMe,
  checkIsSent,
  checkIsPending,
  checkIsFailed,
  isFineResend,
  isFineDelete,
  showMenuTrigger,
} from '../../utils/openChannelUtils';
import { getSenderFromMessage } from '../../utils/openChannelUtils';

interface Props {
  className?: string | Array<string>;
  message: ClientFileMessage;
  disabled: boolean;
  userId: string;
  status: string;
  chainTop: boolean;
  chainBottom: boolean;
  onClick(bool: boolean): void,
  showRemove(bool: boolean): void,
  resendMessage(message: ClientFileMessage): void;
}

export default function OpenchannelThumbnailMessage({
  className,
  message,
  disabled,
  userId,
  status,
  chainTop,
  onClick,
  showRemove,
  resendMessage,
}: Props): JSX.Element {
  const {
    type,
    url,
    localUrl,
    thumbnails,
  } = message;
  const thumbnailUrl = (thumbnails && thumbnails.length > 0 && thumbnails[0].url) || null;
  const { stringSet, dateLocale } = useContext(LocalizationContext);
  const { disableUserProfile, renderUserProfile } = useContext(UserProfileContext);
  const [messageWidth, setMessageWidth] = useState(360);
  const messageRef = useRef(null);
  const contextMenuRef = useRef(null);
  const avatarRef = useRef(null);

  const memorizedThumbnailPlaceHolder = useMemo(() => (type) => ({ style }) => ( // eslint-disable-line
    <div style={style}>
      <Icon
        type={type}
        fillColor={IconColors.ON_BACKGROUND_2}
        width="56px"
        height="56px"
        />
    </div>
  ), []);

  const isByMe = checkIsByMe(message, userId);
  const isMessageSent = checkIsSent(status);
  const isPending = checkIsPending(status);
  const isFailed = checkIsFailed(status);
  const sender = getSenderFromMessage(message);

  useEffect(() => {
    const thumbnailWidth = messageRef?.current?.clientWidth - 80;
    setMessageWidth(thumbnailWidth > 360 ? 360 : thumbnailWidth);
  }, []);

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-openchannel-thumbnail-message',
      ].join(' ')}
      ref={messageRef}
    >
      <div className="sendbird-openchannel-thumbnail-message__left">
        {
          !chainTop && (
            <ContextMenu
              menuTrigger={(toggleDropdown) => (
                <Avatar
                  className="sendbird-openchannel-thumbnail-message__left__avatar"
                  src={sender.profileUrl || ''}
                  ref={avatarRef}
                  width="28px"
                  height="28px"
                  onClick={() => {
                    if (!disableUserProfile) {
                      toggleDropdown();
                    }
                  }}
                />
              )}
              menuItems={(closeDropdown) => (
                <MenuItems
                  parentRef={avatarRef}
                  parentContainRef={avatarRef}
                  closeDropdown={closeDropdown}
                  style={{ paddingTop: 0, paddingBottom: 0 }}
                >
                  {
                    renderUserProfile
                      ? (
                        renderUserProfile({
                          user: sender,
                          close: closeDropdown,
                        })
                      )
                      : (
                        <UserProfile
                          user={sender}
                          onSuccess={closeDropdown}
                          disableMessaging
                        />
                      )
                  }
                </MenuItems>
              )}
            />
          )
        }
      </div>
      <div className="sendbird-openchannel-thumbnail-message__right">
        {
          !chainTop && (
            <div className="sendbird-openchannel-thumbnail-message__right__title">
              <Label
                className="sendbird-openchannel-thumbnail-message__right__title__sender-name"
                type={LabelTypography.CAPTION_2}
                color={isByMe ? LabelColors.SECONDARY_3 : LabelColors.ONBACKGROUND_2}
              >
                {
                  sender && (
                    sender.friendName
                    || sender.nickname
                    || sender.userId
                  )
                }
              </Label>
              <Label
                className="sendbird-openchannel-thumbnail-message__right__title__sent-at"
                type={LabelTypography.CAPTION_3}
                color={LabelColors.ONBACKGROUND_3}
              >
                {format(message.createdAt, 'p', { locale: dateLocale })}
              </Label>
            </div>
          )
        }
        <div className="sendbird-openchannel-thumbnail-message__right__body">
          <div
            className="sendbird-openchannel-thumbnail-message__right__body__wrap"
            role="button"
            onClick={() => {
              if (isMessageSent) {
                onClick(true);
              }
            }}
            onKeyDown={() => {
              if (isMessageSent) {
                onClick(true);
              }
            }}
            tabIndex={0}
          >
            <div className="sendbird-openchannel-thumbnail-message__right__body__wrap__overlay" />
            {
              {
                [SUPPORTING_TYPES.VIDEO]: (
                  (url || localUrl)
                    ? (
                      <div className="sendbird-openchannel-thumbnail-message__right__body__wrap__video" >
                        {
                          (thumbnailUrl)
                            ? (
                              <ImageRenderer
                                className="sendbird-openchannel-thumbnail-message__right__body__wrap__video"
                                url={thumbnailUrl}
                                width={messageWidth}
                                height="270px"
                                alt="image"
                                placeHolder={memorizedThumbnailPlaceHolder(IconTypes.PLAY)}
                              />
                            )
                            : (
                              <video className="sendbird-openchannel-thumbnail-message__right__body__wrap__video__video">
                                <source src={url || localUrl} type={type} />
                              </video>
                            )
                        }
                        <Icon
                          className="sendbird-openchannel-thumbnail-message__right__body__wrap__video__icon"
                          type={IconTypes.PLAY}
                          fillColor={IconColors.ON_BACKGROUND_2}
                          width="56px"
                          height="56px"
                        />
                      </div>
                    )
                    : (
                      <Icon
                        className="sendbird-openchannel-thumbnail-message__right__body__wrap__video--icon"
                        type={IconTypes.PHOTO}
                        fillColor={IconColors.ON_BACKGROUND_2}
                        width="56px"
                        height="56px"
                      />
                    )
                ),
                [SUPPORTING_TYPES.IMAGE]: (
                  (url || localUrl)
                    ? (
                      <ImageRenderer
                        className="sendbird-openchannel-thumbnail-message__right__body__wrap__image"
                        url={thumbnailUrl || url || localUrl}
                        alt="image"
                        width={messageWidth}
                        height="270px"
                        placeHolder={memorizedThumbnailPlaceHolder(IconTypes.PHOTO)}
                      />
                    )
                    : (
                      <Icon
                        className="sendbird-openchannel-thumbnail-message__right__body__wrap__image--icon"
                        type={IconTypes.PHOTO}
                        fillColor={IconColors.ON_BACKGROUND_2}
                        width="56px"
                        height="56px"
                      />
                    )
                ),
                [SUPPORTING_TYPES.UNSUPPORTED]: (
                  <Icon
                    className="sendbird-openchannel-thumbnail-message__right__body__wrap__unknown"
                    type={IconTypes.PHOTO}
                    fillColor={IconColors.ON_BACKGROUND_2}
                    width="56px"
                    height="56px"
                  />
                ),
              }[getSupportingFileType(type)]
            }
          </div>
        </div>
        {
          (isPending || isFailed) && (
            <div className="sendbird-openchannel-thumbnail-message__right__tail">
              {
                isPending && (
                  <Loader
                    width="16px"
                    height="16px"
                  >
                    <Icon
                      className="sendbird-openchannel-thumbnail-message__right__tail__pending"
                      type={IconTypes.SPINNER}
                      fillColor={IconColors.PRIMARY}
                      width="16px"
                      height="16px"
                    />
                  </Loader>
                )
              }
              {
                isFailed && (
                  <Icon
                    className="sendbird-openchannel-thumbnail-message__right__tail__failed"
                    type={IconTypes.ERROR}
                    width="16px"
                    height="16px"
                  />
                )
              }
            </div>
          )
        }
      </div>
      <div
        className="sendbird-openchannel-thumbnail-message__context-menu"
        ref={contextMenuRef}
      >
        <ContextMenu
          menuTrigger={(toggleDropdown) => (
            showMenuTrigger({ message, userId, status }) && (
              <IconButton
                className="sendbird-openchannel-thumbnail-message__context-menu--icon"
                width="32px"
                height="32px"
                onClick={toggleDropdown}
              >
                <Icon
                  type={IconTypes.MORE}
                  fillColor={IconColors.CONTENT_INVERSE}
                  width="24px"
                  height="24px"
                />
              </IconButton>
            )
          )}
          menuItems={(closeDropdown) => (
            <MenuItems
              parentRef={contextMenuRef}
              parentContainRef={contextMenuRef}
              closeDropdown={closeDropdown}
              openLeft
            >
              {
                isFineResend({ message, userId, status }) && (
                  <MenuItem
                    onClick={() => {
                      resendMessage(message);
                      closeDropdown();
                    }}
                  >
                    {stringSet.CONTEXT_MENU_DROPDOWN__RESEND}
                  </MenuItem>
                )
              }
              {
                isFineDelete({ message, userId, status }) && (
                  <MenuItem
                    onClick={() => {
                      if (disabled) {
                        return;
                      }
                      showRemove(true);
                      closeDropdown();
                    }}
                  >
                    {stringSet.CONTEXT_MENU_DROPDOWN__DELETE}
                  </MenuItem>
                )
              }
            </MenuItems>
          )}
        />
      </div>
    </div>
  );
}
