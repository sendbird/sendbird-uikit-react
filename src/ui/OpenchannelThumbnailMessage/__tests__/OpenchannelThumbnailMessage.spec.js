import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import mockMessage, { mockMessageVideo, getMockMessage, getMockMessageWithVideo } from '../dummyData.mock';

import OpenchannelThumbnailMessage from "../index";

// mock date-fns to avoid problems from snapshot timestamping
// between testing in different locations
// ideally we want to mock date-fns globally - needs more research
jest.mock('date-fns/format', () => () => ('mock-date'));

describe('OpenchannelThumbnailMessage', () => {
  it('should render for image thumbnail message', function() {
    const component = mount(
      <OpenchannelThumbnailMessage
        message={mockMessage}
        status="succeeded"
        userId="hoon102"
      />
    );

    expect(
      component.find('.sendbird-openchannel-thumbnail-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__left__avatar').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sender-name').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sent-at').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__overlay').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__image').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__context-menu--icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should render for empty url image thumbnail message', function() {
    const mockMessage = getMockMessage();
    mockMessage.url = '';
    const component = mount(
      <OpenchannelThumbnailMessage
        message={mockMessage}
        status="succeeded"
        userId="hoon102"
      />
    );

    expect(
      component.find('.sendbird-openchannel-thumbnail-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__left__avatar').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sender-name').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sent-at').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__overlay').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__image').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__image--icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__context-menu--icon').exists()
    ).toBe(true);expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(false);
  });


  it('should render for video thumbnail message', function() {
    const component = mount(
      <OpenchannelThumbnailMessage
        message={mockMessageVideo}
        status="succeeded"
        userId="hoon102"
      />
    );

    expect(
      component.find('.sendbird-openchannel-thumbnail-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__left__avatar').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sender-name').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sent-at').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__overlay').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video__video').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video__icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__context-menu--icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should render for empty url video thumbnail message', function() {
    const mockMessage = getMockMessageWithVideo();
    mockMessage.url = '';
    const component = mount(
      <OpenchannelThumbnailMessage
        message={mockMessage}
        status="succeeded"
        userId="hoon102"
      />
    );

    expect(
      component.find('.sendbird-openchannel-thumbnail-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__left__avatar').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sender-name').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sent-at').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__overlay').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video__video').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video__icon').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__video--icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__context-menu--icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should not render avatar and title when chainTop is true', function() {
    const component = mount(
      <OpenchannelThumbnailMessage
        message={mockMessage}
        status="succeeded"
        userId="hoon102"
        chainTop
      />
    );

    expect(
      component.find('.sendbird-openchannel-thumbnail-message').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__left__avatar').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sender-name').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__title__sent-at').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__overlay').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__image').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__body__wrap__image--icon').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__context-menu--icon').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should render pending icon if status is pending', function() {
    const message = getMockMessageWithVideo((message) => {
      message.isResendable = () => true;
      message.sendingStatus = 'pending';
    });
    const component = mount(
      <OpenchannelThumbnailMessage
        message={message}
        userId={message.sender.userId}
      />
    );
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(false);
  });

  it('should render failed icon if status is failed', function() {
    const message = getMockMessageWithVideo((message) => {
      message.isResendable = () => true;
      message.sendingStatus = 'failed';
    });
    const component = mount(
      <OpenchannelThumbnailMessage
        message={message}
        userId={message.sender.userId}
      />
    );
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__pending').exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-openchannel-thumbnail-message__right__tail__failed').exists()
    ).toBe(true);
  });

  it('should do a snapshot test of the OpenchannelThumbnailMessage DOM', function() {
    const component = renderer.create(
      <OpenchannelThumbnailMessage
        message={mockMessage}
        status="succeeded"
        userId="hoon102"
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
