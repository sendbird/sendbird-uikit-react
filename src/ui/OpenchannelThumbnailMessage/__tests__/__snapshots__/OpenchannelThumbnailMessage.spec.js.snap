// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/OpenchannelThumbnailMessage should do a snapshot test of the OpenchannelThumbnailMessage DOM 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-openchannel-thumbnail-message"
  >
    <div
      class="sendbird-openchannel-thumbnail-message__left"
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <div
          class="sendbird-openchannel-thumbnail-message__left__avatar sendbird-avatar"
          role="button"
          style="height: 28px; width: 28px; z-index: 0; left: 0px;"
          tabindex="0"
        >
          <div
            class="sendbird-avatar-img sendbird-image-renderer"
            style="width: 100%; min-width: min(400px, 28px); max-width: 400px;"
          >
            <div
              class="sendbird-image-renderer__image"
              style="width: 100%; min-width: min(400px, 28px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_15_512px.png);"
            />
            <img
              alt=""
              class="sendbird-image-renderer__hidden-image-loader"
              src="https://static.sendbird.com/sample/profiles/profile_15_512px.png"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sendbird-openchannel-thumbnail-message__right"
    >
      <div
        class="sendbird-openchannel-thumbnail-message__right__title"
      >
        <span
          class="sendbird-openchannel-thumbnail-message__right__title__sender-name sendbird-label sendbird-label--caption-2 sendbird-label--color-onbackground-2"
        >
          hoon102
        </span>
        <span
          class="sendbird-openchannel-thumbnail-message__right__title__sent-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-3"
        >
          mock-date
        </span>
      </div>
      <div
        class="sendbird-openchannel-thumbnail-message__right__body"
      >
        <div
          class="sendbird-openchannel-thumbnail-message__right__body__wrap"
          role="button"
          tabindex="0"
        >
          <div
            class="sendbird-openchannel-thumbnail-message__right__body__wrap__overlay"
          />
          <div
            class="sendbird-openchannel-thumbnail-message__right__body__wrap__image sendbird-image-renderer"
            style="width: 100%; min-width: min(400px, -80px); max-width: 400px;"
          >
            <div
              style="width: 100%; min-width: min(400px, -80px); max-width: 400px; position: absolute; display: flex; justify-content: center; align-items: center;"
            >
              <div
                class=" sendbird-icon sendbird-icon-photo sendbird-icon-color--on-background-2"
                role="button"
                style="width: 56px; min-width: 56px; height: 56px; min-height: 56px;"
                tabindex="0"
              >
                <test-file-stub />
              </div>
            </div>
            <div
              class="sendbird-image-renderer__image"
              style="width: 100%; min-width: min(400px, -80px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://sendbird-staging.s3.amazonaws.com/26FA710D-A7F6-4483-ABEA-1F07A07A112E/upload/n/e49e0a4111d4445a8840bd51f3c9d933.png);"
            />
            <img
              alt="image"
              class="sendbird-image-renderer__hidden-image-loader"
              src="https://sendbird-staging.s3.amazonaws.com/26FA710D-A7F6-4483-ABEA-1F07A07A112E/upload/n/e49e0a4111d4445a8840bd51f3c9d933.png"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sendbird-openchannel-thumbnail-message__context-menu"
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <button
          class="sendbird-openchannel-thumbnail-message__context-menu--icon sendbird-iconbutton "
          style="height: 32px; width: 32px;"
          type="button"
        >
          <span
            class="sendbird-iconbutton__inner"
          >
            <div
              class=" sendbird-icon sendbird-icon-more sendbird-icon-color--content-inverse"
              role="button"
              style="width: 24px; min-width: 24px; height: 24px; min-height: 24px;"
              tabindex="0"
            >
              <test-file-stub />
            </div>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
